src/main.rs

```diff
@@ -246,70 +302,95 @@ async fn main() -> Result<()> {
             addr,
             auth_token,
             key,
+            rpc_port,
         } => {
-            tokio::select! {
-                biased;
-                res = provide_interactive(path, addr, auth_token, key, cli.keylog) => {
-                    res
-                }
-                _ = tokio::signal::ctrl_c() => {
-                    println!("\nShutting down provider...");
-                    Ok(())
-                }
+            let provider = provide(addr, auth_token, key, cli.keylog, rpc_port).await?;
+            let controller = provider.controller();
+            let mut ticket = provider.ticket(Hash::from([0u8; 32]));
+
+            // task that will add data to the provider, either from a file or from stdin
+            let fut = tokio::spawn(async move {
+                let (path, tmp_path) = if let Some(path) = path {
+                    (path, None)
+                } else {
+                    // Store STDIN content into a temporary file
+                    let (file, path) = tempfile::NamedTempFile::new()?.into_parts();
+                    let mut file = tokio::fs::File::from_std(file);
+                    let path_buf = path.to_path_buf();
+                    // Copy from stdin to the file, until EOF
+                    tokio::io::copy(&mut tokio::io::stdin(), &mut file).await?;
+                    // return the TempPath to keep it alive
+                    (path_buf, Some(path))
+                };
+                // tell the provider to add the data
+                let ProvideResponse { hash } = controller.rpc(ProvideRequest { path }).await??;
+                ticket.hash = hash;
+                println!("Collection: {}", Blake3Cid::new(hash));
+                println!("All-in-one ticket: {}", ticket);
+                // println!("{}", response.hash);
+                // let (db, hash) = provider::create_collection(sources).await?;
+                // println!("Collection: {}\n", Blake3Cid::new(hash));
+                // let mut total_size = 0;
+                // for (_, path, size) in db.blobs() {
+                //     total_size += size;
+                //     println!("- {}: {}", path.display(), HumanBytes(size));
+                // }
+                // println!("Total: {}", HumanBytes(total_size));
+                // println!();
+                // return all the things that need to live even after this task
+                // println!("All-in-one ticket: {}", provider.ticket(hash));
+                anyhow::Ok(tmp_path)
+            });
+
+            tokio::signal::ctrl_c().await?;
+            println!("Shutting down provider...");
+            provider.shutdown();
+            provider.await?;
+            fut.abort();
+            drop(fut);
```

Comment:
why is this needed?

Reply:
the future holds the temp file. When it gets dropped, the temp file gets deleted. Goddamn temp file... Added a comment.